### PR TITLE
fetch pending emails

### PR DESCRIPTION
**Pull request by ahalberkamp**
Right now the emailprocessor allways queries the 100 oldest emailobjects independent from their Status. After that the processor iterate throught all the emailobjects, and only send emails for the Objects, when they have the State "pending" or if the flag of the cronjob "processAll" it set to true .

The Problem is:

When the 100 "oldest Email" are in the state "Error", the emailprocessor allways fetches this 100 Emails and doesn´t process them unless the Flag "processAll" is set to true.

To solve this problems 2 changes was Implemented:

1. The check, if emailobects are in state "pending" , is directly made within the "CustomObjectQuery" and not while iteration over the returned emailobjects.

2. The emailprocessor now query all "pending" emailobjects using the "queryAll" feature

### CODE REVIEW

File: src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java [hunk 1]
```diff
@@ -63,42 +63,26 @@ public EmailProcessor() {
...
-                LOG.error(String.format("[%s] An unknown error occurred", tenantConfig.getProjectKey()),
-                    exception);
-                client.close();
-                return Statistics.ofError(tenantConfig.getProjectKey());
-            });
+        CustomObjectQuery<JsonNode> query = CustomObjectQuery.ofJsonNode().byContainer(CONTAINER_ID);
+        if (!tenantConfig.isProcessAll()) {
+            query = query.plusPredicates(QueryPredicate.of("value(status=\"" + STATUS_PENDING + "\")"));
+        }
+        query = query.withSort(s -> s.createdAt().sort().asc());
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java]: Maybe add a comment why you sort ascendingly above?

File: src/test/java/com/commercetools/emailprocessor/email/EmailProcessorTest.java [hunk 2]
```diff
@@ -103,32 +107,32 @@ public void processEmail_pendingEmailAvailable_shouldProcessEmails() throws Exce
     @Test
     public void processEmail_processAllFlagIsSet_shouldProcessAllEmails() throws Exception {
         customObjects.add(createCustomObject("1", STATUS_PENDING, RESPONSE_CODE_SUCCESS));
-        customObjects.add(createCustomObject("2", STATUS_PENDING, Statistics.RESPONSE_ERROR_TEMP));
-        customObjects.add(createCustomObject("3", EMAIL_STATUS_ERROR, Statistics.RESPONSE_ERROR_TEMP));
+        customObjects.add(createCustomObject("2", STATUS_PENDING, RESPONSE_ERROR_TEMP));
+        customObjects.add(createCustomObject("3", EMAIL_STATUS_ERROR, RESPONSE_ERROR_TEMP));
         customObjects.add(createCustomObject("4", STATUS_PENDING, Statistics.RESPONSE_ERROR_PERMANENT));
-        tenantConfiguration.setClient(mockSphereClient(customObjects));
+        tenantConfiguration.setClient(mockSphereClient(customObjects, false));
```
Comment (heshamMassoud) [inline src/test/java/com/commercetools/emailprocessor/email/EmailProcessorTest.java]: but the method name contains `..processAllFlagIsSet..`, and you are passing `false`?

File: src/integration-test/java/com/commercetools/emailprocessor/integration/commons/jobs/EmailJobIT.java [hunk 3]
```diff
@@ -88,12 +88,28 @@ public void process_withASuccessfulEmail_shouldReturnNoError() {
...
+        }
+        configuration.getTenants().get(0).setEndpointUrl(HTTPBIN_DOMAIN + Statistics.RESPONSE_CODE_SUCCESS);
+        List<Statistics> statistics = EmailJob.process(configuration).toCompletableFuture().join();
+        assertThat(statistics).isNotEmpty();
+        Statistics statistic = statistics.get(0);
+        assertThat(statistic.getProcessed()).isEqualTo(100);
+        assertThat(statistic.getSentSuccessfully()).isEqualTo(100);
+        assertThat(statistic.getTemporaryErrors()).isEqualTo(0);
+        assertThat(statistic.getPermanentErrors()).isEqualTo(0);
+    }
```
Comment (heshamMassoud) [inline src/integration-test/java/com/commercetools/emailprocessor/integration/commons/jobs/EmailJobIT.java]: Maybe it makes sense to add ITs for asserting that the email objects have all been processed after running the process (that's in the case if processAll for example) and one for the case (when not processAll is set and there are errored emails for example). So the cases to test are:

`process_withOnlyPendingEmailObjectsAndProcessAllIsFalse_shouldProcessAll`
`process_withOnlyPendingEmailObjectsAndProcessAllIsTrue_shouldProcessAll`
`process_withMixedStateEmailObjectsAndProcessAllIsFalse_shouldNotProcessAll`
`process_withMixedStateEmailObjectsAndProcessAllIsTrue_shouldProcessAll`

These ITs will assert that the feature that you added works now

File: src/test/java/com/commercetools/emailprocessor/model/StatisticsTest.java [hunk 4]
```diff
@@ -43,7 +42,6 @@ public void ofError_isCalledForTenant_shouldShowCorrectStatstic() throws Excepti
         assertEquals(statistics.getPermanentErrors(), 0);
         assertEquals(statistics.getTemporaryErrors(), 0);
         assertEquals(statistics.getProcessed(), 0);
-        assertEquals(statistics.getNotProcessed(), 0);
```
Comment (heshamMassoud) [inline src/test/java/com/commercetools/emailprocessor/model/StatisticsTest.java]: Why are all these asserts removed in this unit test?

File: src/test/java/com/commercetools/emailprocessor/model/StatisticsTest.java [hunk 5]
```diff
@@ -25,14 +25,13 @@ public void update_differentResponseCodesAreProvided_shouldShowCorrectStatistic(
         statistics.update(Statistics.RESPONSE_CODE_SUCCESS);
         statistics.update(Statistics.RESPONSE_ERROR_TEMP);
         statistics.update(Statistics.RESPONSE_ERROR_PERMANENT);
-        statistics.update(0);
-        statistics.update(0);
+
```
Comment (heshamMassoud) [inline src/test/java/com/commercetools/emailprocessor/model/StatisticsTest.java]: Why are these lines removed?

File: src/test/java/com/commercetools/emailprocessor/model/StatisticsTest.java (hunk 4 above)
Comment (ahalberkamp) [reply src/test/java/com/commercetools/emailprocessor/model/StatisticsTest.java]: because the attribute "notprocessed" doesn`t exist anymore in the statistic, because this state could not be "reached" in the current implementation

File: src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java [hunk 6]
```diff
@@ -63,42 +63,26 @@ public EmailProcessor() {
...
+        Statistics statistics = new Statistics(tenantConfig.getProjectKey());
+
+        Function<CustomObject<JsonNode>, CompletableFuture<Integer>> callBack = customObject ->
+                callApiEndpoint(customObject.getId(), tenantConfig);
+
+        return QueryExecutionUtils.queryAll(client, query, callBack)
+                .thenApply(result -> result.stream()
+                        .map(httpResponse -> httpResponse.thenAccept(code -> statistics.update(code)))
```
[the remark this answers is not in the document]
Comment (ahalberkamp) [reply src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java]: ef888bf43d7ed77f2e4c4635c7207fe08f26d682

File: src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java [hunk 7]
```diff
@@ -63,42 +63,26 @@ public EmailProcessor() {
...
+                        .collect(toList()))
+                .thenApply(futures -> {
+                    futures.forEach(CompletableFuture::join);
```
[the remark this answers is not in the document]
Comment (ahalberkamp) [reply src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java]: 04e113b2c53a5165504e82ebf8290a58d132ec47

File: src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java (hunk 1 above)
Comment (ahalberkamp) [reply src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java]: 04e113b2c53a5165504e82ebf8290a58d132ec47

File: src/test/java/com/commercetools/emailprocessor/model/StatisticsTest.java (hunk 5 above)
Comment (ahalberkamp) [reply src/test/java/com/commercetools/emailprocessor/model/StatisticsTest.java]: Because the state "0" doesn`t exist anymore

File: src/integration-test/java/com/commercetools/emailprocessor/integration/commons/jobs/EmailJobIT.java (hunk 3 above)
Comment (ahalberkamp) [reply src/integration-test/java/com/commercetools/emailprocessor/integration/commons/jobs/EmailJobIT.java]: 879b5c3

File: src/test/java/com/commercetools/emailprocessor/email/EmailProcessorTest.java (hunk 2 above)
Comment (ahalberkamp) [reply src/test/java/com/commercetools/emailprocessor/email/EmailProcessorTest.java]: 5dfb7016e3f7338c560a9116fca9e69fdd1186ac

File: src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java [hunk 8]
```diff
@@ -63,42 +61,22 @@ public EmailProcessor() {
...
-                        .orElseGet(() -> {
-                            statistics.update(Statistics.RESPONSE_IGNORED);
-                            return CompletableFuture.completedFuture(null);
-                        }))
-                    .collect(toList());
+        CustomObjectQuery<JsonNode> query = CustomObjectQuery.ofJsonNode().byContainer(CONTAINER_ID);
+        if (!tenantConfig.isProcessAll()) {
+            query = query.plusPredicates(QueryPredicate.of("value(status=\"" + STATUS_PENDING + "\")"));
+        }
+        //sorting the emailobject by creationt ime, to ensure that the emails are delivered in the correct chronically
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java]: @ahalberkamp Some typos. Maybe ->

`// We sort the email object by creation time, to ensure that the emails are delivered in the correct chronological order they were sent with`

File: src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java [hunk 9]
```diff
@@ -63,42 +61,22 @@ public EmailProcessor() {
...
-                allOf(allTenants.toArray(new CompletableFuture[0])).join();
+        final Consumer<CustomObject<JsonNode>> pageConsumer = customObject ->
+                callApiEndpoint(customObject.getId(), tenantConfig)
+                        .thenAccept(statistics::update)
+                        .toCompletableFuture().join();
 
-                client.close();
-                return statistics;
-            })
-            .exceptionally(exception -> {
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java]: Is there a reason why the exception handeling is removed?  also you don't need to close the client anymore?

File: src/test/java/com/commercetools/emailprocessor/model/StatisticsTest.java (hunk 4 above)
Comment (heshamMassoud) [reply src/test/java/com/commercetools/emailprocessor/model/StatisticsTest.java]: Why is this not possible? can you please clarify this change in logic as it's not described any where in the PR description?

Comment (ahalberkamp) [reply src/test/java/com/commercetools/emailprocessor/model/StatisticsTest.java]: It is described in the pr-description:

The check, if emailobects are in state "pending" , is directly made within the "CustomObjectQuery" and not while iteration over the returned emailobjects.

So the Result of the customobject query doesn`t has any object which shoudn´t be  processed

File: src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java (hunk 9 above)
Comment (ahalberkamp) [reply src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java]: 01cd4477f8b2ae17befad4040e9bbd7fe2233762

File: src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java (hunk 8 above)
Comment (ahalberkamp) [reply src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java]: 7c4c7f4e01e02c348f5b58790a081fd42e252ece

File: src/test/java/com/commercetools/emailprocessor/model/StatisticsTest.java (hunk 4 above)
Comment (heshamMassoud) [reply src/test/java/com/commercetools/emailprocessor/model/StatisticsTest.java]: ok

File: src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java [hunk 10]
```diff
@@ -56,55 +55,48 @@ public EmailProcessor() {
...
-                return Statistics.ofError(tenantConfig.getProjectKey());
-            });
+        final SphereClient client = tenantConfig.getSphereClient();
+        CustomObjectQuery<JsonNode> query = CustomObjectQuery.ofJsonNode().byContainer(CONTAINER_ID);
+        if (!tenantConfig.isProcessAll()) {
+            query = query.plusPredicates(QueryPredicate.of("value(status=\"" + STATUS_PENDING + "\")"));
+        }
+        // We sort the email object by creation time, to ensure that the emails are delivered in the correct
+        // chronological order they were sent with.
+        query = query.withSort(s -> s.createdAt().sort().asc());
```
Comment (butenkor) [inline src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java]: @heshamMassoud @ahalberkamp We should use here pagination by UUID and not custom sort and offset (due to know limitations). If it is not supported by CTP JVM SDK yet than we need to create an issue here and address it once non offset based pagination is possible over SDK.

File: src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java [hunk 11]
```diff
@@ -114,15 +106,15 @@ public EmailProcessor() {
...
-                final String encyptedCustomerId = blowFish(customObjectId, tenantConfiguration.getEncryptionKey(),
-                    Cipher.ENCRYPT_MODE);
-                params.add(new BasicNameValuePair(PARAM_EMAIL_ID, encyptedCustomerId));
+                final String encryptedCustomerId = blowFish(customObjectId, tenantConfiguration.getEncryptionKey(),
+                        Cipher.ENCRYPT_MODE);
+                params.add(new BasicNameValuePair(PARAM_EMAIL_ID, encryptedCustomerId));
                 params.add(new BasicNameValuePair(PARAM_TENANT_ID, tenantConfiguration.getProjectKey()));
                 httpPost.setEntity(new UrlEncodedFormEntity(params, Charset.defaultCharset()));
                 return doPost(HttpClients.createDefault(), httpPost, tenantConfiguration.getProjectKey());
-            } catch (Exception excepton) {
```
Comment (butenkor) [inline src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java]: Nice transformers like misspelling. Sounds like one of decepticons members :)

File: src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java [hunk 12]
```diff
@@ -56,55 +55,48 @@ public EmailProcessor() {
     }
 
     /**
-     * Iterate through all Email objects and triggers the webhook for each pending email object.
+     * Iterate through all Email objects and triggers the web hook for each pending email object.
```
[the remark this answers is not in the document]
Comment (ahalberkamp) [reply src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java]: 7dcee0e845efb2e4595b421c27836b231afbe9bf

File: src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java [hunk 13]
```diff
@@ -56,55 +55,48 @@ public EmailProcessor() {
...
-            .exceptionally(exception -> {
-                LOG.error(String.format("[%s] An unknown error occurred", tenantConfig.getProjectKey()),
-                    exception);
-                client.close();
-                return Statistics.ofError(tenantConfig.getProjectKey());
-            });
+        final SphereClient client = tenantConfig.getSphereClient();
+        CustomObjectQuery<JsonNode> query = CustomObjectQuery.ofJsonNode().byContainer(CONTAINER_ID);
+        if (!tenantConfig.isProcessAll()) {
+            query = query.plusPredicates(QueryPredicate.of("value(status=\"" + STATUS_PENDING + "\")"));
```
[the remark this answers is not in the document]
Comment (ahalberkamp) [reply src/main/java/com/commercetools/emailprocessor/email/EmailProcessor.java]: a40454dae9a2deeb2cadd464c0028acecabab4d7

File: src/test/java/com/commercetools/emailprocessor/email/EmailProcessorTest.java [hunk 14]
```diff
@@ -192,18 +211,22 @@ public void doPost_httpPostReturnStatusLine_shouldReturnCorrectMessage() throws
         assertEquals(RESPONSE_CODE_SUCCESS, result);
     }
 
-    private SphereClient mockSphereClient(final List<CustomObject<JsonNode>> customObjects) {
+    private SphereClient mockSphereClient(final List<CustomObject<JsonNode>> customObjects, boolean processAll) {
```
[the remark this answers is not in the document]
Comment (ahalberkamp) [reply src/test/java/com/commercetools/emailprocessor/email/EmailProcessorTest.java]: c4fde14

File: src/test/java/com/commercetools/emailprocessor/email/EmailProcessorTest.java [hunk 15]
```diff
@@ -115,20 +120,34 @@ public void processEmail_processAllFlagIsSet_shouldProcessAllEmails() throws Exc
...
     }
 
+    @Test
+    public void processEmail_ClientThrowsException_shouldHandleException() {
+        SphereClient client = mock(SphereClient.class);
+        when(client.execute(any(CustomObjectQuery.class)))
+                .thenReturn(exceptionallyCompletedFuture(new Exception("anyError")));
+        tenantConfiguration.setClient(client);
+        tenantConfiguration.setProcessAll(true);
+        Statistics statistic = emailProcessor.processEmails(tenantConfiguration).toCompletableFuture().join();
```
[the remark this answers is not in the document]
Comment (ahalberkamp) [reply src/test/java/com/commercetools/emailprocessor/email/EmailProcessorTest.java]: bf259a1

File: src/test/java/com/commercetools/emailprocessor/email/EmailProcessorTest.java [hunk 16]
```diff
@@ -192,18 +211,22 @@ public void doPost_httpPostReturnStatusLine_shouldReturnCorrectMessage() throws
         assertEquals(RESPONSE_CODE_SUCCESS, result);
     }
 
-    private SphereClient mockSphereClient(final List<CustomObject<JsonNode>> customObjects) {
+    private SphereClient mockSphereClient(final List<CustomObject<JsonNode>> customObjects, boolean processAll) {
         SphereClient client = mock(SphereClient.class);
-        final PagedQueryResult<CustomObject<JsonNode>> queryResult = PagedQueryResult.of(customObjects);
+        final PagedQueryResult<CustomObject<JsonNode>> queryResult = PagedQueryResult.of(customObjects.stream()
+                .filter(co -> StringUtils.equals(co.getValue().get("status").asText(), STATUS_PENDING)
+                        || processAll)
```
[the remark this answers is not in the document]
Comment (ahalberkamp) [reply src/test/java/com/commercetools/emailprocessor/email/EmailProcessorTest.java]: d29cf0d

**Comment by heshamMassoud** [comment]
@ahalberkamp Regarding a7c25f8,

1. I wouldn't name the gradle script `junit-tests.gradle`. Since we use junit for both unit and integration tests. Do you mean more `unit-tests.gradle`?
2. It seems you also moved the integration test configuration there: https://github.com/commercetools/commercetools-email-retry-processor/commit/a7c25f8351505d26cf78e8bf7e3f152c6bbcf5b1#diff-7d80c59300db0a883d66909a48c16330R3. I guess you copied it by mistake.
3. Maybe it makes sense instead of having a configuration in multiple gradle scripts, just have all configurations together in a separate script `configurations.gradle`, WDYT?
